### ISSUE TITLE
hi，can you wirte a description of how to implement NNCLR?

### ISSUE THREAD

**Issue by HBU-Lin-Li** [lightly-ai/lightly#414]
hi，can you wirte a description of how to implement NNCLR?
Thanks.

**Comment by MalteEbner** [comment]
We have already implemented NNCLR and added a description how to use it:
https://github.com/lightly-ai/lightly/blob/8972b58cb1c445a7591e5394be5b548a066fa044/lightly/models/nnclr.py#L87-L93

**Comment by IgorSusmelj** [comment]
Hi @HBU-Lin-Li , do you mean something like this?
```python
import torchvision
import torch.nn as nn
from lightly.models import NNCLR
from lightly.loss import NTXentLoss
from lightly.models.modules import NNMemoryBankModule

resnet = torchvision.models.resnet18()
backbone = nn.Sequential(
    *list(resnet.children())[:-1],
    nn.AdaptiveAvgPool2d(1),
)

# NNCLR
model = NNCLR(backbone)
criterion = NTXentLoss()

nn_replacer = NNMemoryBankModule(size=2 ** 16)

# forward pass
(z0, p0), (z1, p1) = model(x0, x1)
z0 = nn_replacer(z0.detach(), update=False)
z1 = nn_replacer(z1.detach(), update=True)
loss = 0.5 * (criterion(z0, p1) + criterion(z1, p0))
```

**Comment by MalteEbner** [comment]
Hi @HBU-Lin-Li,
I assume, this closes this issue.
If it doesn't, feel free to reopen it.

**Comment by ramdhan1989** [comment]
> Hi @HBU-Lin-Li , do you mean something like this?
> 
> ```python
> import torchvision
> import torch.nn as nn
> from lightly.models import NNCLR
> from lightly.loss import NTXentLoss
> from lightly.models.modules import NNMemoryBankModule
> 
> resnet = torchvision.models.resnet18()
> backbone = nn.Sequential(
>     *list(resnet.children())[:-1],
>     nn.AdaptiveAvgPool2d(1),
> )
> 
> # NNCLR
> model = NNCLR(backbone)
> criterion = NTXentLoss()
> 
> nn_replacer = NNMemoryBankModule(size=2 ** 16)
> 
> # forward pass
> (z0, p0), (z1, p1) = model(x0, x1)
> z0 = nn_replacer(z0.detach(), update=False)
> z1 = nn_replacer(z1.detach(), update=True)
> loss = 0.5 * (criterion(z0, p1) + criterion(z1, p0))
> ```

Hi @IgorSusmelj ,
would you mind providing complete code until training step ? I don't understand what are x0 and x1 parameter in model(x0, x1) under forward pass comment ?

thanks

**Comment by philippmwirth** [comment]
Hi @ramdhan1989, thanks for the question!

Please take a look at our [tutorial on SimSiam](https://docs.lightly.ai/tutorials/package/tutorial_simsiam_esa.html). To use NNCLR instead of SimSiam, you can simply replace the model initialization and the forward pass by what @IgorSusmelj wrote above.

As you can see in the inner for loop of the training loop, the `x0` and `x1` are the two augmentations of the input images:

![grafik](https://user-images.githubusercontent.com/65946090/123599128-8f88d200-d7f5-11eb-99e3-49bd12fdfeb5.png)


I hope this helps! Let me know if there are any further questions. We will definitely consider adding a full example but I can't make any promises on the timeline.

Cheers,

**Comment by ramdhan1989** [comment]
Thanks @philippmwirth. I modified it as follow. is this correct ?
```
model = torchvision.models.alexnet(pretrained=False)
model = model.features
model_ssl = models.NNCLR(model, num_ftrs=256)
criterion = loss.NTXentLoss()
nn_replacer = NNMemoryBankModule(size=2**16)

optimizer = torch.optim.SGD(model_ssl.parameters(), lr=1e-2,momentum=0.9, weight_decay=5e-4)
running_loss = 0
train_losses, test_losses = [], []

for e in range(max_epochs):
    start = time.time()
    for (x0, x1), _, _ in dataloader:
        (z0, p0), (z1, p1) = model_ssl(x0, x1)
        z0 = nn_replacer(z0.detach(), update=False)
        z1 = nn_replacer(z1.detach(), update=True)
        loss = 0.5 * (criterion(z0, p1) + criterion(z1, p0))
        loss.backward()

        optimizer.step()
        optimizer.zero_grad()
        
    print(f'[Epoch {e:3d}] '
        f'Loss = {loss:.2f} | '
        )
```

**Comment by philippmwirth** [comment]
Yes, this looks correct

**Comment by HBU-Lin-Li** [comment]
Hi,
z0 and z1 are the selected nearest neighbors ? And their shapes are related to be the batch size?
Thanks.

**Comment by HBU-Lin-Li** [comment]
Hi, @IgorSusmelj ,thanks for your help.
Now, I have another question : z0 and z1 are the selected nearest neighbors ? And their shapes are related to be the batch size?
Such that  
        (z0, p0), (z1, p1) = model_ssl(x0, x1)
        z0 = nn_replacer(z0.detach(), update=False)
        z1 = nn_replacer(z1.detach(), update=True),

and final z0 and z1's shapes are respective batch_size * d?

**Comment by HBU-Lin-Li** [comment]
Hi, @IgorSusmelj ,thanks for your help.
Now, I have another question : z0 and z1 are the selected nearest neighbors ? And their shapes are related to be the batch size?
Such that  
        (z0, p0), (z1, p1) = model_ssl(x0, x1)
        z0 = nn_replacer(z0.detach(), update=False)
        z1 = nn_replacer(z1.detach(), update=True),

and final z0 and z1's shapes are respective batch_size * d?

**Comment by philippmwirth** [comment]
> 
> 
> Hi, @IgorSusmelj ,thanks for your help.
> Now, I have another question : z0 and z1 are the selected nearest neighbors ? And their shapes are related to be the batch size?
> Such that
> (z0, p0), (z1, p1) = model_ssl(x0, x1)
> z0 = nn_replacer(z0.detach(), update=False)
> z1 = nn_replacer(z1.detach(), update=True),
> 
> and final z0 and z1's shapes are respective batch_size * d?

Yes, `z0` and `z1` are the selected nearest neighbors and they have the same shape as before which means `batch_size x d`

**Comment by HBU-Lin-Li** [comment]
In NNCLR,the number of NN can be changed. In this code baseline, can we change the number of the chosed NN,such as 1?

**Comment by philippmwirth** [comment]
The number of nearest neighbors is fixed at 1 per sample. If you want to change this you would need to inherit from `lightly.models.modules.NNMemoryBankModule` and adapt the code at

https://github.com/lightly-ai/lightly/blob/32efe8cb5493d4936d5a7fbc58425ae26380fdef/lightly/models/modules/nn_memory_bank.py#L59

**Comment by HBU-Lin-Li** [comment]
Thanks for your patience.

**Comment by philippmwirth** [comment]
Closing this due to inactivity.
